### ISSUE TITLE
Nexus panicked when there are unhandled instance state update exceptions

### ISSUE THREAD

**Issue by askfongjojo** [oxidecomputer/omicron#6090]
I found a nexus core file on rack2. Here are the log lines just before it crashed: 
```
BRM42220057 # head -215310 /pool/ext/b4a71d3d-1ecd-418a-9a52-8d118f82082b/crypt/debug/oxz_nexus_20b100d0-84c3-4119-aa9b-0c632b0b6a3a/oxide-nexus:default.log.1720884601 | tail -10 | looker
15:27:36.278Z WARN 20b100d0-84c3-4119-aa9b-0c632b0b6a3a (ServerContext): failed to update instance from sled agent
    error = InternalError { internal_message: "unexpected database error: root: memory budget exceeded: 122880 bytes requested, 134137872 currently allocated, 134217728 bytes in budget" }
    file = nexus/src/app/instance.rs:2180
    instance_id = 073db4e9-0822-4e84-b128-aa3f86a615ef
    propolis_id = 74c289d6-5886-4aa1-9609-32e35cb63b8b
15:27:36.278Z INFO 20b100d0-84c3-4119-aa9b-0c632b0b6a3a (dropshot_internal): request completed
    error_message_external = Internal Server Error
    error_message_internal = unexpected database error: root: memory budget exceeded: 122880 bytes requested, 134137872 currently allocated, 134217728 bytes in budget
    file = /home/build/.cargo/git/checkouts/dropshot-a4a923d29dccc492/9fef396/dropshot/src/server.rs:866
    latency_us = 15251
    local_addr = [fd00:1122:3344:104::3]:12221
    method = PUT
    remote_addr = [fd00:1122:3344:104::1]:63747
    req_id = 9d1f119e-108d-4413-b20e-9d2624e87c93
    response_code = 500
    uri = /instances/073db4e9-0822-4e84-b128-aa3f86a615ef
thread 'tokio-runtime-worker' panicked at nexus/db-queries/src/db/sec_store.rs:141:14:
called `Result::unwrap()` on an `Err` value: InternalError { internal_message: "database error (kind = Unknown): scan with start key /Table/297/1/\"\\xb9\\\\?2\\x1d\\xfbBA\\xa0m\\xad\\x82\\xf9^\\xb9\\xbf\"/0: root: memory budget exceeded: 133120 bytes requested, 134137872 currently allocated, 134217728 bytes in budget\nHINT: Consider increasing --max-sql-memory startup parameter.\n" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[ Jul 13 15:27:38 Stopping because all processes in service exited. ]
```
The root cause of the database error is #5904 which is pending a fix. We probably don't want such errors to cause nexus to panic and lead to other failure modes.

The nexus core file is being uploaded to catacomb:/staff/core/omicron-6090.

**Comment by smklein** [comment]
I think this is #2416 biting us again, since the panic is from `nexus/db-queries/src/db/sec_store.rs:141`

**Comment by davepacheco** [comment]
I thought we had fixed that, but it looks like #5390 only fixed the `record_event()` path.  I think we can probably apply similar behavior to the `saga_update()` path.

Looking at both this and the `record_event()` path: we might need to consider that if there's a transient error, the change may have successfully been made.  (Imagine that CockroachDB successfully handles the query and commits the transaction and then there's a network blip that causes us to get an ECONNRESET or something.)  I think we should probably treat certain errors here (like a conflict) as a success.

We should also think through the case that the underlying state has changed.  e.g., suppose we invoke `saga_update` to set the state to "Running", this succeeds under the hood but generates a transient error, then at some point later the saga becomes "done", but this retry loop sets it back to "running".  I think this should be impossible because we don't support failover today, and won't in the near future without also having expunged Nexus, so there's nobody else who could have changed this in the meantime.  Still, I'm not sure the code has thought about this possibility and we should at least document why it's safe to ignore conflicts.

**Comment by sunshowers** [comment]
This should be addressed in #6113 -- @askfongjojo should I close this out or do you want to get this into dogfood and monitor this further?

**Comment by askfongjojo** [comment]
Thank you for addressing this issue! I'll check it when dogfood has the fix.
